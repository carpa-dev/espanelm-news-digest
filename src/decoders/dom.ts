import { JSDOM } from 'jsdom';
import { Result } from 'true-myth';

export interface DOM {
  querySelector(q: string): null | Element;
  querySelectorAll(q: string): NodeListOf<Element>;
}

export function DOMFromText(s: string): Result<DOM, { reason: string }> {
  try {
    return Result.ok(new JSDOMImpl(s));
  } catch (error) {
    return Result.err({ reason: error });
  }
}

// JSDOM does some runtime to guarantee
// it's running inside a class
// therefore we need this wrapper
export class JSDOMImpl implements DOM {
  _dom: JSDOM;

  constructor(html: string) {
    this._dom = new JSDOM(html);
  }

  querySelector(q: string): Element | null {
    return this._dom.window.document.querySelector(q);
  }

  querySelectorAll(q: string): NodeListOf<Element> {
    return this._dom.window.document.querySelectorAll(q);
  }
}
