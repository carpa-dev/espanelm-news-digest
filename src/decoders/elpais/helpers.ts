import { DOM } from '../dom';
import { Maybe } from 'true-myth';

export function array<K>(dom: DOM, q: string): Maybe<K[]> {
  const body = Array.from(dom.querySelectorAll(q))
    .map(b => b.textContent || '')
    .filter(b => b.length > 1);

  if (body.length) {
    return Maybe.of(body);
  }
  return Maybe.nothing();
}
