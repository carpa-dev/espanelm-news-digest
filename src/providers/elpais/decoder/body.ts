import { DOM } from '../../../decoders/dom';
import { Maybe } from 'true-myth';

export function decodeBody(dom: DOM): Maybe<string[]> {
  return method1(dom).orElse(method2.bind(null, dom));
}

function method1(dom: DOM): Maybe<string[]> {
  return array(dom, '[itemprop="articleBody"] p');
}

function method2(dom: DOM): Maybe<string[]> {
  return array(dom, '.article_body p');
}

function array(dom: DOM, q: string): Maybe<string[]> {
  const body = Array.from(dom.querySelectorAll(q))
    .map(b => b.textContent || '')
    .filter(b => b.length > 1);

  if (body.length) {
    return Maybe.of(body);
  }
  return Maybe.nothing();
}
