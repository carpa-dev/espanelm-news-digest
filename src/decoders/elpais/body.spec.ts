import { loadFixtureOrFail } from '../../../test/helpers';
import { DOM, DOMFromText } from '../dom';
import { decodeBody } from './body';
import { Result } from 'true-myth';

describe('ElPais Body Decoder', () => {
  const domFromFixture = async (f: string): Promise<DOM> => {
    const fixture = await loadFixtureOrFail(f);

    return Result.unsafelyUnwrap(DOMFromText(fixture));
  };

  const unwrappedBody = (dom: DOM): string[] =>
    decodeBody(dom).unsafelyUnwrap();

  it('should work with method 1', async () => {
    const dom = await domFromFixture('article.html');

    expect(unwrappedBody(dom)).toMatchSnapshot();
  });

  it('should work with method 2', async () => {
    const dom = await domFromFixture('articles/1565345150_208544');

    expect(unwrappedBody(dom)).toMatchSnapshot();
  });

  it('should return Nothing if its unable to decode', async () => {
    const dom = DOMFromText('').unsafelyUnwrap();
    const body = decodeBody(dom);
    expect(body.isNothing()).toBe(true);
  });
});
