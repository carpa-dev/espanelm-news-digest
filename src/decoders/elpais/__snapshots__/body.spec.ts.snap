// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElPais Body Decoder should work with method 1 1`] = `
Array [
  "O abra√ßo entre duas crian√ßas nas ruas de Nova York viralizou nas redes. O v√≠deo flagrou Maxwell e Finnegan, ambos de dois anos, correndo um para o outro com os bra√ßos abertos ap√≥s se ver na rua. O pai de Maxwell, Michael Cisneros, que publicou a cena no domingo passado, disse que os menores se conhecem desde um ano e desde ent√£o s√£o insepar√°veis. ‚ÄúSempre est√£o muito empolgados de se ver, mesmo que estejam separados por apenas um ou dois dias‚Äù, comentou Cisneros ao Huffington Post.",
  "‚ÄúSimplesmente correram um para o outro, e peguei meu celular o mais r√°pido que pude para gravar. S√£o muito lindos juntos \\", disse o pai a WPIX, um canal de televis√£o de Nova York. ‚ÄúCom todo o racismo e √≥dio, acho que √© um v√≠deo realmente emocionante‚Äù. Inclusive, a representante democrata norte-americana Alexandria Ocasio-Cortez compartilhou a cena em suas redes sociais.",
  "ü•∞ https://t.co/RDaEZkRxvC",
]
`;

exports[`ElPais Body Decoder should work with method 2 1`] = `
Array [
  "Para explicar onde mora, Olimpio Guajajara, de 44 anos, leva a m√£o ao bolso. Saca o celular, abre um aplicativo e mostra um quadrado na tela; dentro dele, chamas indicam dois inc√™ndios ativos. S√£o imagens obtidas via sat√©lite. Monitoramento com tecnologia ponta. Ele est√° na frente mais dura da batalha contra o desmatamento, porque se encarrega da vigil√¢ncia no terreno, que exige destreza e √© muito mais perigosa. Cuida de sua terra como antes fizeram seus pais, av√≥s, bisav√¥s e demais antepassados. Mas al√©m dos inc√™ndios de sempre, muito mais vorazes neste inverno, os inimigos se multiplicaram. Agora est√£o mais poderosos. O ativista lidera os guardi√£es do territ√≥rio ind√≠gena Arariboia, 123 membros da tribo dos guajajara organizados para proteger o que nos mapas √© uma ilha verde de vegeta√ß√£o em plena Amaz√¥nia, mas rodeada por zonas desmatadas.",
  "O desmatamento na Amaz√¥nia brasileira cresceu 222% em agosto deste ano com rela√ß√£o ao mesmo m√™s de 2018, segundo o sistema do INPE (Instituto Nacional de Pesquisas Espaciais) que alerta os fiscais. Esse aumento e o discurso do presidente Jair Bolsonaro, que menospreza a mudan√ßa clim√°tica, odeia as ONGs e quer autorizar atividades mineradoras em terras ind√≠genas, colocaram a destrui√ß√£o das florestas no centro do debate pol√≠tico local e internacional. Algo que ficou claro no G7 organizado pelo presidente Emmanuel Macron na Fran√ßa. Para Bolsonaro, trata-se de pura inger√™ncia, porque ele considera que a Amaz√¥nia diz respeito apenas aos brasileiros.",
  "Ind√≠genas como Olimpio Guajajara s√£o quem melhor conhece e quem usa de forma mais s√°bia e sustent√°vel as florestas tropicais, uma esp√©cie de ar condicionado natural para refrescar um planeta cada vez mais quente. Sua presen√ßa √© crucial para frear o aquecimento global. ‚ÄúProtegemos o pulm√£o da terra‚Äù, diz o guardi√£o na sede do Instituto Socioambiental de S√£o Paulo, onde esteve no final de julho para denunciar as amea√ßas que espreitam a mata. ‚ÄúAntes faz√≠amos [a fiscaliza√ß√£o] a p√©, agora nos doaram v√°rias motos‚Äù, conta. Usam armas? ‚ÄúN√£o, n√£o temos armas de fogo, temos arcos e flechas.‚Äù Os sat√©lites s√£o complementados com armas ancestrais nesta guerra do s√©culo XXI contra a crise clim√°tica.",
  "Quando perguntado sobre a extens√£o do territ√≥rio, √© extremamente preciso: ‚Äúabrange 413.288,47 hectares‚Äù, responde, salientando at√© a √∫ltima cifra. Com quase o triplo da √°rea do munic√≠pio de S√£o Paulo, fica no extremo leste da Amaz√¥nia, no Estado do Maranh√£o. L√° ele vive com a esposa e seis dos seus oito filhos. Os dois mais velhos, um militar e uma pedagoga, migraram para a cidade. A terra Ararib√≥ia √© o lar de 5.300 pessoas, pertencentes √†s tribos dos guajajara e aos seus vizinhos aw√° guaj√°, que somam apenas 60 ou 70 indiv√≠duos. Pouco se sabe desse outro grupo, porque est√£o entre os poucos ind√≠genas n√£o contatados no Brasil. Quando os guajajara revelaram aos brancos que naquele pedacinho de terra havia outra tribo, inicialmente n√£o acreditaram neles. Mas sim, l√° estavam, e l√° continuam, os aw√° guaj√°. Foram filmados pelos guajajara, mas n√£o demonstram nenhum interesse em interagir com os estranhos √† sua comunidade. O guardi√£o considera que proteger esses ind√≠genas e seu modo de vida √© tamb√©m parte de sua miss√£o. Explica isso de maneira pausada, em um portugu√™s que aprendeu depois de sua l√≠ngua materna, o tenetehara.",
  "‚ÄúA terra est√° sendo invadida por madeireiros‚Äù, afirma, ‚Äúmas tamb√©m por traficantes de drogas, ca√ßadores, pescadores‚Ä¶‚Äù. Sua outra grande queixa √© o pouco que o Estado se deixa ver naquelas terras para fiscalizar, para assegurar que o lugar √© protegido como manda a lei. E quando detectam um invasor, n√£o √© f√°cil dar o alarme. L√° n√£o funcionam os telefones nem a Internet. Nem os r√°dios.",
  "Conta que depois de tr√™s anos de gest√µes ‚Äîviver na mata exige enormes doses de paci√™ncia ‚Äî, finalmente conseguiram incluir alguns representantes institucionais em uma opera√ß√£o de julho. Um batalh√£o de oito policiais ambientais e tr√™s funcion√°rios da Funai se uniu a 14 guardi√£es guajajara em uma expedi√ß√£o em busca de madeireiros furtivos. ‚ÄúN√≥s os levamos ao ponto exato‚Äù, diz, orgulhoso de sua equipe. No primeiro lugar onde chegaram, pilharam-nos com as m√£os na massa. Revela que eram dois √≠ndios e tr√™s brancos. A pol√≠cia confiscou o trator, abriu um inqu√©rito, mas conv√©m lembrar que os crimes ambientais t√™m uma taxa de impunidade alt√≠ssima no Brasil. Durante a patrulha, descobriram que os madeireiros tinham aberto cem quil√¥metros de estrada com ramais. No segundo ponto, s√≥ restavam os troncos derrubados. Com uma mistura de lamento e indigna√ß√£o, ele conta que ap√≥s nove dias os policiais e fiscais se foram. O Estado, que tinha prometido 30 dias de patrulhas, se recolhia de novo. Os guardi√£es tinham nove potenciais cen√°rios localizados de delitos. O descomunal tamanho do Brasil e o car√°ter remoto de muitos dos lugares mais vulner√°veis dificultam enormemente a fiscaliza√ß√£o. Desde que Bolsonaro assumiu o cargo, as multas ambientais desabaram.",
  "Embora tenha vivido com outras comunidades durante a adolesc√™ncia, Olimpio Guajajara retornou √† sua terra. ‚Äú√â nossa m√£e, dela vivemos, nos d√° comida, nos d√° sa√∫de‚Ä¶ Por isso estamos empenhados em defender a floresta para toda a humanidade‚Äù. O modo de vida na terra Ararib√≥ia ‚Äîoficialmente reconhecida como uma reserva ind√≠gena em 1990, depois de uma batalha judicial de duas d√©cadas‚Äî √© diferente do de seus antepassados. Por um lado, agora h√° escola, mas por outro tamb√©m surgem casos de c√¢ncer e diabetes. J√° n√£o se alimentam s√≥ de ca√ßa ‚Äîtamb√©m comem sardinha, frango e feijoada.",
  "Guajajara salienta que os guardi√£es da terra Arariboia precisam de refor√ßos para proteger a floresta contra o presidente Bolsonaro, a quem considera ‚Äúum anti-ind√≠gena declarado‚Äù, os madeireiros e a longa lista de inimigos que pretendem que desistam de proteger seus 413.288,47 hectares da Amaz√¥nia.",
]
`;
